// baileys-redis-auth/lib/src/Example/logger-pino.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
const pino_1 = __importDefault(require("pino"));
const package_json_1 = __importDefault(require("#/../package.json"));
const pretty = {
    level: 'info',
    target: 'pino-pretty',
    options: {
        colorize: true,
        translateTime: 'SYS:isoDateTime',
        ignore: 'pid,hostname',
    },
};
const transports = pino_1.default.transport({
    targets: [pretty],
});
const options = {
    level: process.env.PINO_LOG_LEVEL || 'info',
    formatters: {
        bindings: (bindings) => {
            return {
                pid: bindings.pid,
                host: bindings.hostname,
                app: package_json_1.default.name,
                v: package_json_1.default.version,
                node_version: process.version,
            };
        },
        level: (label, number) => {
            return { level: number, label: label.toUpperCase() };
        },
    },
    timestamp: pino_1.default.stdTimeFunctions.isoTime,
};
exports.logger = (0, pino_1.default)(options, transports);
